File: stockflow-backend/src/main/java/com/stockflow/backend/service/BrandService.java
package com.stockflow.backend.service;

import com.stockflow.backend.dto.brand.BrandRequest;
import com.stockflow.backend.dto.brand.BrandResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * BrandService — Service interface defining the Brand module contract.
 */
public interface BrandService {

    /**
     * Create a new brand.
     * Throws DuplicateResourceException if the name already exists.
     */
    BrandResponse createBrand(BrandRequest request);

    /**
     * Retrieve a brand by its ID.
     * Throws BrandNotFoundException if not found.
     */
    BrandResponse getBrandById(Long id);

    /**
     * Retrieve all active brands (non-paginated).
     */
    List<BrandResponse> getAllBrands();

    /**
     * Retrieve a paginated and sorted list of all active brands.
     */
    Page<BrandResponse> getAllBrandsPaginated(Pageable pageable);

    /**
     * Update an existing brand by ID.
     */
    BrandResponse updateBrand(Long id, BrandRequest request);

    /**
     * Soft-delete a brand by setting active = false.
     */
    void deleteBrand(Long id);

    /**
     * Search active brands by keyword.
     */
    Page<BrandResponse> searchBrands(String keyword, Pageable pageable);
}

File: stockflow-frontend/src/features/brand/services/brandService.js
import api from '@/services/api';
import { API } from '@/constants/api';

/**
 * brandService
 * GET /api/brands/page    → paginated (params: page, size, sort, direction)
 * GET /api/brands/search  → search (params: keyword, page, size, sort, direction)
 * GET /api/brands         → full list (for dropdowns)
 *
 * NOTE: Brand/Supplier/Warehouse use 'sort' (not 'sortBy') and '/page' subpath.
 */
export const brandService = {
  getPaginated: async (page = 0, size = 10, sortBy = 'name', direction = 'ASC') => {
    const res = await api.get(`${API.BRANDS}/page`, {
      params: { page, size, sort: sortBy, direction },
    });
    return res.data;
  },

  getAllList: async () => {
    const res = await api.get(API.BRANDS);
    return res.data;
  },

  search: async (keyword, page = 0, size = 10, sortBy = 'name', direction = 'ASC') => {
    const res = await api.get(`${API.BRANDS}/search`, {
      params: { keyword, page, size, sort: sortBy, direction },
    });
    return res.data;
  },

  create: async (data) => {
    const res = await api.post(API.BRANDS, data);
    return res.data;
  },

  update: async (id, data) => {
    const res = await api.put(`${API.BRANDS}/${id}`, data);
    return res.data;
  },

  delete: async (id) => {
    const res = await api.delete(`${API.BRANDS}/${id}`);
    return res.data;
  },
};
